Sort generated properties by their keys to get deterministic output.

Zaid-Ajaj opened the pull request:
Fixes #109

Frassle commented on pkg/convert/tf.go:
```diff
@@ -993,6 +993,10 @@ func convertObjectConsExpr(state *convertState, inBlock bool, scopes *scopes,
 			Value: valueTokens,
 		})
 	}
+	// Sort the items by name to make the output deterministic
```
Try and sort by HCL source position. That keeps things in the declaration order from the original.

Zaid-Ajaj replied:
Okay that makes more sense

Zaid-Ajaj replied:
Done ✅

justinvp commented:
The sorting makes the naming deterministic (what Anton was seeing in #110)?

Zaid-Ajaj commented:
@Frassle I think I have fixed the non-determinism in naming.  When we rewrite the traversal of invokes/data sources we use `scopes.getOrAddPulumiName` effectively changing the state during traversal without distinguishing between bound vs. unbound invokes. I simplified it so that we don't mutate the scope when we encounter an unbound invoke/data source and instead just renaming it. Same for unbound resources not using `scopes.getOrAddPulumiName` 

The renaming rules for unbound stuff is simple. for data sources `data.{data_source_token}.{local_name}.{rest}` becomes `{localName}{DateSourceToken}.{rest}` 

For unbound resources `{resource_type}.{local_name}.{rest}` becomes `{localName}{resourceType}.{rest}` (similar to what it was doing before)

Frassle commented on pkg/convert/tf.go:
```diff
@@ -952,6 +952,12 @@ func convertObjectConsExpr(state *convertState, inBlock bool, scopes *scopes,
 ) hclwrite.Tokens {
 	items := []hclwrite.ObjectAttrTokens{}
+
+	// Sort the items by line number so that we can deterministically order them in the output
```
I don't think these sorts are actually needed, pretty sure Items is already going to be in declaration order.

Frassle replied:
Again I'm pretty sure Items is already sorted, if so this is just extra code for no benefit.

Frassle commented on pkg/convert/testdata/programs/partial_attribute_lookup/pcl/main.pp:
```diff
@@ -1,4 +1,4 @@
 resource "example" "simple:index:resource" {
-  inputOne = exampledataunknownDataSource.attr
-  inputTwo = exampleunknownResourceType.list[0]
+  inputOne = exampleUnknownDataSource.attr
```
These names look better

Frassle commented on pkg/convert/tf.go:
```diff
@@ -1299,10 +1314,12 @@ func rewriteTraversal(
 					newTraversal = append(newTraversal, rewriteRelativeTraversal(scopes, "", traversal[1:])...)
 				} else {
-					// We don't know what this is, so lets assume it's an unknown resource (we shouldn't ever have unknown locals)
-					newName = scopes.getOrAddPulumiName(path, "", camelCaseName(root.Name))
```
So I think the concern of _not_ using getOrAddPulumiName is that if the picked name here conflicts with some other bound or unbound name that's already in use we'll get a conflict in the PCL.
Feels like we could probably write a test case to cover that, and then fix this up.

Frassle reviewed (commented):
I don't think this helps with Antons case at all but the unbound names do look nicer if we can get them without conflicts with other things.

justinvp commented:
Note: https://github.com/pulumi/pulumi-converter-terraform/pull/116 should address the example test failure.

Zaid-Ajaj commented on pkg/convert/tf.go:
```diff
@@ -1195,9 +1207,19 @@ func rewriteTraversal(
-			newName := scopes.getOrAddPulumiName(path, "", "data"+suffix)
-			newTraversal = append(newTraversal, hcl.TraverseRoot{Name: newName})
-			newTraversal = append(newTraversal, rewriteRelativeTraversal(scopes, path, traversal[3:])...)
+			if scopes.lookup(path) != "" {
```
Changed it to use the result from `lookup` ✅ it is technically the same as `camelCaseName(secondAttr.name)` but better get it from the lookup table

Zaid-Ajaj commented on pkg/convert/tf.go:
```diff
@@ -1589,6 +1614,10 @@ type bodyAttrTokens struct {
 
 func tokensForObject(ts []bodyAttrTokens) hclwrite.Tokens {
 	attrs := make([]hclwrite.ObjectAttrTokens, 0, len(ts))
+	// sort body attribute tokens by name to get a deterministic output
```
Did these sorts as a just in case because wasn't sure what was and what wasn't sorted. Removed them now
